### PR TITLE
Implemented various Gen. 8 abilities

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
This PR provides the effect of ~~the first 6~~ many abilities defined from the Gen. 8 Games.

**EDIT**: As of right now, the abilities introduced by this PR are the following:

Intrepid Sword, Dauntless Shield, Libero, Cotton Down, Ball Fetch, Propeller Tail, Stalwart, Steam Engine, Sand Spit, Perish Body, Punk Rock, Ice Scales and Power Spot.

I do not intend to work in any other ability for the time being, so this PR is good to go minus potential corrections, as far as I am concerned.

## **Discord contact info**
Lunos#4026

### CODE REVIEW

**Review by ghoulslash — DISMISSED** [review]

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Some more issues from testing:
- Player's mon gets Intrepid Sword boost from foe's ability
    - In double battles, any activation (even from your first slot mon) seems to boost your second mon
- Same issue with Dauntless Shield
- Cotton Down on second round of activation from a multi-strike move seems to softlock
    - Ability Pop-up should be modified to stay open once #1244 is merged
- No ability pop-up for Ball Fetch

Also, you can implement Stalwart as well since it's the same as Propeller Tail.

File: src/battle_script_commands.c
Comment (LOuroboros) [reply src/battle_script_commands.c:11996]: The Master Ball doesn't need to be taken into account.

It won't fail in a wild battle, and if lost by misusing it on a trainer battle, a Pokémon with Ball Fetch does not get it back, just like in Sword and Shield :eyes:

Comment (LOuroboros) [reply src/battle_script_commands.c:11996]: Also, that check is not redundant because the one at L11932 affects gLastUsedItem, not gLastUsedBall.

I did try to make use of that originally, but I couldn't. That is the reason why I had to add gLastUsedBall :eyes:

File: src/battle_main.c
Comment (LOuroboros) [reply src/battle_main.c:232]: May I ask what is the reasoning behind, for future reference? :eyes:

**Comment by LOuroboros** [comment]
> Some more issues from testing:
> 
>     * Player's mon gets Intrepid Sword boost from foe's ability
>       * In double battles, any activation (even from your first slot mon) seems to boost your second mon
>     * Same issue with Dauntless Shield
>     * Cotton Down on second round of activation from a multi-strike move seems to softlock
>       * Ability Pop-up should be modified to stay open once #1244 is merged
>     * No ability pop-up for Ball Fetch
> 
> Also, you can implement Stalwart as well since it's the same as Propeller Tail.

I figured I should write up a quick update.

I have addressed every single one of those *(with the help of Blurose, Ghoulslash and Churchofcage)* **[in a separate branch](https://github.com/LOuroboros/pokeemerald/commits/gen8_abilities_BE_dev)**, except for the ability pop up behavior during Cotton Down. I'm still waiting for #1244 to be merged so I can address that :eyes:

**Comment by LOuroboros** [comment]
> 
> 
> > Some more issues from testing:
> > ```
> > * Player's mon gets Intrepid Sword boost from foe's ability
> >   * In double battles, any activation (even from your first slot mon) seems to boost your second mon
> > * Same issue with Dauntless Shield
> > * Cotton Down on second round of activation from a multi-strike move seems to softlock
> >   * Ability Pop-up should be modified to stay open once #1244 is merged
> > * No ability pop-up for Ball Fetch
> > ```
> > 
> > 
> > Also, you can implement Stalwart as well since it's the same as Propeller Tail.
> 
> I figured I should write up a quick update.
> 
> I have addressed every single one of those _(with the help of Blurose, Ghoulslash and Churchofcage)_ **[in a separate branch](https://github.com/LOuroboros/pokeemerald/commits/gen8_abilities_BE_dev)**, except for the ability pop up behavior during Cotton Down. I'm still waiting for #1244 to be merged so I can address that 👀

#1244 was finally merged, and I was corrected the AbilityPopUp usage in Cotton Down's BattleScript.

I kept working on **[that separate branch](https://github.com/LOuroboros/pokeemerald/commits/gen8_abilities_BE_dev)** and on top of the fixes to the issues raised in this PR and adding Stalwart, I implemented some more abilities. Namely:

-Steam Engine
-Sand Spit
-Perish Body
-Punk Rock
-Ice Scales
-Power Spot *(though I feel like this one needs some extra testing.)*

I'm ready to push everything to the branch that I submitted this PR from if there are no objections :eyes:

**Comment by ultima-soul** [comment]
Yeah, you can push those changes to this PR if you want, but obviously it'll take more time for testing.

Alternatively, you could only add changes to the abilities that were originally in this PR + Stalwart and submit the other abilities as a new PR. That would speed things up for this particular PR.

**Comment by LOuroboros** [comment]
> 
> 
> Yeah, you can push those changes to this PR if you want, but obviously it'll take more time for testing.
> 
> Alternatively, you could only add changes to the abilities that were originally in this PR + Stalwart and submit the other abilities as a new PR. That would speed things up for this particular PR.

But it would probably still take quite some time for the new PR to be approved, so eh.

**Review by ultima-soul — COMMENTED** [review]
Does Punk Rock really need its own flag? I thought it only would've needed the Sound flag.

**Comment by LOuroboros** [comment]
> 
> 
> Does Punk Rock really need its own flag? I thought it only would've needed the Sound flag.

As we discussed in Discord, it certainly doesn't. Punk Rock simply needs to check for the Sound flag and leave status moves out of it.

Will fix in a moment.

**Comment by LOuroboros** [comment]
Ready for re-review :eyes:

File: src/battle_util.c
Comment (ultima-soul) [inline src/battle_util.c:7237]: I think the check for being a Status move is redundant because only damaging attacks will have Damage Calc commands in their move effect script.

Comment (LOuroboros) [reply src/battle_util.c:7237]: That makes sense to me. I'll remove it :eyes:

**Comment by LOuroboros** [comment]
Ready for re-re-review :eyes:

**Comment by ultima-soul** [comment]
Could you merge in the latest Battle Engine? This PR currently doesn't build from clean.

**Comment by LOuroboros** [comment]
> 
> 
> Could you merge in the latest Battle Engine? This PR currently doesn't build from clean.

Done :eyes:

File: data/battle_scripts_1.s
Comment (ghoulslash) [inline data/battle_scripts_1.s:6774]: This should be `STRINGID_PKMNSXWHIPPEDUPSANDSTORM`, imo

**Review by ghoulslash — APPROVED** [review]

File: src/battle_util.c
Comment (ghoulslash) [inline src/battle_util.c:7105]: according to bulbapedia, this boosts power not attack, in which case it should go under `Battery` in `CalcMoveBasePowerAfterModifiers`

Comment (LOuroboros) [reply src/battle_util.c]: Done :eyes:

**Review by ultima-soul — CHANGES_REQUESTED** [review]
The error with gBattleResults.catchAttempts access is still there, which I believe we've discussed how to fix on Discord.

**Comment by LOuroboros** [comment]
Ready for... uh... I don't know, re-re-re-re-re-review or something 🤣

**Review by ultima-soul — DISMISSED** [review]

File: src/battle_util.c
Comment (ghoulslash) [inline src/battle_util.c:4647]: should just be `SET_STATCHANGER(STAT_SPEED, 6, FALSE)` to raise it by 6 stages

File: src/battle_util.c
Comment (ghoulslash) [inline src/battle_util.c:4649]: Should be `BattleScript_TargetAbilityStatRaiseOnMoveEnd`

**Comment by LOuroboros** [comment]
Ready for yet another re-filled review :eyes:

**Review by ghoulslash — APPROVED** [review]

**Review by ghoulslash — APPROVED** [review]

**Review by ultima-soul — APPROVED** [review]

**Pull request merged**
